src/dmd/cond.d

```diff
@@ -362,13 +373,66 @@ extern (C++) final class StaticForeach : RootObject
         auto catass = new CatAssignExp(aloc, new IdentifierExp(aloc, idres), res[1]);
         s2.push(createForeach(aloc, pparams[1], new ExpStatement(aloc, catass)));
         s2.push(new ReturnStatement(aloc, new IdentifierExp(aloc, idres)));
-        auto aggr = wrapAndCall(aloc, new CompoundStatement(aloc, s2));
-        sc = sc.startCTFE();
-        aggr = aggr.expressionSemantic(sc);
-        aggr = resolveProperties(sc, aggr);
-        sc = sc.endCTFE();
-        aggr = aggr.optimize(WANTvalue);
-        aggr = aggr.ctfeInterpret();
+
+        Expression aggr;
+        Type indexty;
+
+        if (rangefe && (indexty = ety.typeSemantic(aloc, sc)).isintegral())
+        {
+            // min = (upr - lwr)
+            Expression min = new MinExp(aloc, rangefe.upr, rangefe.lwr);
+            min = min.expressionSemantic(sc);
+            min = min.optimize(WANTvalue);
+
+            // gt = (upr > lwr)
+            Expression gt = new CmpExp(TOK.greaterThan, aloc, rangefe.upr, rangefe.lwr);
+            gt = gt.expressionSemantic(sc);
+            gt = gt.optimize(WANTvalue);
```

Comment:
If rangefe's lwr and upr expressions are guaranteed to be constant at this point,  would it be better to shortcut optimize and call the constfold routines directly, or do the comparisons between the toInteger() result? Thus avoid needing to allocate for a result that is immediately discarded.

https://github.com/dlang/dmd/blob/5b827474c014cade5c1a1a25d27de843ffdca4c5/src/dmd/optimize.d#L1091

Probably not much to gain from doing so, as this does not look like a frequently called path.

Reply:
Though I have a nagging memory that IntRange fits this purpose as well for ensuring that an expression doesn't exceed the bounds of a type.

Reply:
Just tried using the functions directly but expressionSemantic can wrap a IntegerExp with a CastExp and optimize() take care of that so it's not worth it.

Reply:
> Though I have a nagging memory that IntRange fits this purpose as well for ensuring that an expression doesn't exceed the bounds of a type.

I saw that in `Foreach` semantics but just coded my own version (which shows the same old behavior) but now I'm studying the `Foreach` to `For` transformation and it seems I just need this piece of code from statementsem.d:
```
SignExtendedNumber lower = getIntRange(fs.lwr).imin;
SignExtendedNumber upper = getIntRange(fs.upr).imax;
if (lower <= upper)
{
    fs.key.range = new IntRange(lower, upper);
}
```

Reply:
Yeah, something to the effect of:

```
auto lwrRange = getIntRange(rangefe.lwr);
auto uprRange = getIntRange(rangefe.upr);

if (lwrRange.imin <= uprRange.imax)
{
    auto bounds = IntRange.fromType(Type.tsize_t);

    if (!bounds.contains(uprRange - lwrRange))
        error(aloc, "`static %s` iterations cannot exceed `size_t.max`, ...);
}
```

Reply:
And you can test it with:
```
/*
REQUIRED_ARGS: -m32
TEST_OUTPUT:
---
fail_compilation/sfe.d(10): Error: static foreach iterations cannot exceed size_t.max, range expression is (0 .. 4294967296L)
---
*/
void main()
{
    static foreach (i; 0 .. cast(long)(size_t.max) + 1) {}
}
```